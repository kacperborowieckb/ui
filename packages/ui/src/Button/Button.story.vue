<script lang="ts" setup>
import { logEvent } from 'histoire/client'
import Button from './Button.vue'
</script>

<template>
  <Story>
    <Variant title="Solid">
      <Button variant="solid" label="Hello World" @click="logEvent('Click', $event)" />
    </Variant>
    <Variant title="Outline">
      <Button variant="outline" label="Hello World" @click="logEvent('Click', $event)" />
    </Variant>
  </Story>
</template>
