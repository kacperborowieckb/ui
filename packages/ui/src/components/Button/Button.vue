<template>
  <button ref="buttonTemplateRef" :class="buttonVariants({ variant })">
    <component :is="buttonContentWrapper" v-bind="buttonContentWrapperProps">
      <span :key="Math.random()" class="flex items-center justify-center gap-2">
        <slot />
      </span>
    </component>
  </button>
</template>

<script lang="ts">
import { tv } from 'tailwind-variants'
import { onMounted, Transition, useTemplateRef } from 'vue'
import type { VariantProps } from 'tailwind-variants'

const buttonVariants = tv({
  base: 'focus-visible:ring-offset-background focus-visible:ring-ring min-w-(--button-width) min-h-(--button-height) relative inline-flex max-w-fit cursor-pointer items-center justify-center overflow-hidden rounded-md px-4 py-1 ring-offset-1 transition-colors duration-200 ease-out focus-visible:outline-none focus-visible:ring-2 disabled:pointer-events-none disabled:opacity-75',
  variants: {
    variant: {
      solid: 'bg-primary text-primary-foreground hover:bg-primary/90 active:bg-primary/85 active:text-primary-foreground/95',
      gradient: 'bg-linear-[170deg] from-primary/85 to-primary hover:from-primary/80 hover:to-primary active:from-primary/75 active:to-primary/95 text-primary-foreground active:text-primary-foreground/95',
      outline: 'border-input text-foreground hover:bg-accent/75 hover:text-accent-foreground active:bg-accent active:text-accent-foreground/80 border',
    },
  },
})

type ButtonVariantProps = VariantProps<typeof buttonVariants>

export interface ButtonProps {
  variant?: ButtonVariantProps['variant']
  animated?: boolean
}
</script>

<script setup lang="ts">
const props = withDefaults(defineProps<ButtonProps>(), {
  variant: 'solid',
})

const buttonRef = useTemplateRef('buttonTemplateRef')

const buttonContentWrapper = props.animated ? Transition : 'span'

const buttonContentWrapperProps = props.animated ? { name: 'shift' } : {}

onMounted(() => {
  if (!props.animated)
    return

  const buttonWidth = buttonRef.value?.offsetWidth
  const buttonHeight = buttonRef.value?.offsetHeight

  buttonRef.value?.style.setProperty('--button-width', `${buttonWidth}px`)
  buttonRef.value?.style.setProperty('--button-height', `${buttonHeight}px`)
})
</script>

<style scoped>
.shift-enter-active {
  transition: all 0.3s var(--ease-out);
}

.shift-leave-active {
  transition: all 0.3s var(--ease-out);
}

.shift-enter-from {
  position: absolute;
  transform: translateY(-16px);
  filter: blur(4px);
  opacity: 0;
}

.shift-leave-to {
  position: absolute;
  filter: blur(4px);
  opacity: 0;
  transform: translateY(16px);
}
</style>
