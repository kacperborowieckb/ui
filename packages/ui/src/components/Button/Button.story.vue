<template>
  <Story>
    <Variant title="Solid">
      <div class="flex gap-4">
        <Button variant="solid" @click="logEvent('Click', $event)">
          Submit
        </Button>
        <Button
          disabled
          variant="solid"
          @click="logEvent('Click', $event)"
        >
          Submit
        </Button>
      </div>
    </Variant>
    <Variant title="Outline">
      <div class="flex gap-4">
        <Button variant="outline" @click="logEvent('Click', $event)">
          Submit
        </Button>
        <Button
          disabled
          variant="outline"
          @click="logEvent('Click', $event)"
        >
          Submit
        </Button>
      </div>
    </Variant>
    <Variant title="Gradient">
      <div class="flex gap-4">
        <Button variant="gradient" @click="logEvent('Click', $event)">
          Submit
        </Button>
        <Button
          disabled
          variant="gradient"
          @click="logEvent('Click', $event)"
        >
          Submit
        </Button>
      </div>
    </Variant>
    <Variant title="Animated">
      <Button
        animated
        variant="gradient"
        @click="startButtonAnimation"
      >
        <template v-if="isLoading">
          <Spinner />
        </template>
        <template v-else>
          Submit
        </template>
      </Button>
    </Variant>
  </Story>
</template>

<script lang="ts" setup>
import { logEvent } from 'histoire/client'
import { ref } from 'vue'

import Spinner from '../Spinner/Spinner.vue'
import Button from './Button.vue'

const isLoading = ref(false)

function startButtonAnimation() {
  isLoading.value = true

  setTimeout(() => {
    isLoading.value = false
  }, 2000)
}
</script>
