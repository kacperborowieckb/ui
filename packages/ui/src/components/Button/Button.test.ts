import { shallowMount } from '@vue/test-utils'

import Button from './Button.vue'

describe('button', () => {
  it('should emit a click event when clicked', async () => {
    const wrapper = shallowMount(Button)

    await wrapper.trigger('click')

    expect(wrapper.emitted()).toHaveProperty('click')
  })

  it('should correctly render solid button', () => {
    const wrapper = shallowMount(Button, {
      slots: {
        default: 'Submit',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should correctly render gradient button', () => {
    const wrapper = shallowMount(Button, {
      props: {
        variant: 'gradient',
      },
      slots: {
        default: 'Submit',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should correctly render outline button', () => {
    const wrapper = shallowMount(Button, {
      props: {
        variant: 'outline',
      },
      slots: {
        default: 'Submit',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should correctly render animated button', () => {
    const wrapper = shallowMount(Button, {
      props: {
        animated: true,
      },
      slots: {
        default: 'Submit',
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
