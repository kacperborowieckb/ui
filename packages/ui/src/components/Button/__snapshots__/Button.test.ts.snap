// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`button > outline > should render an outline button 1`] = `"<button class="border-2 px-2.5 py-1.5 text-sm font-semibold focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 border-blue-500 text-blue-500 hover:bg-blue-50 active:bg-blue-100">Hello</button>"`;

exports[`button > solid > should render a solid button 1`] = `"<button class="border-2 px-2.5 py-1.5 text-sm font-semibold focus:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 border-transparent bg-blue-500 text-white hover:bg-blue-600 active:bg-blue-700">Hello</button>"`;
